src/passes/function_passes/CheckPointPass.cpp

```diff
@@ -0,0 +1,157 @@
+#include <map>
+#include <stdint.h>
+#include <llvm/Pass.h>
+#include <llvm/IR/Intrinsics.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/BasicBlock.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/Module.h>
+#include <llvm/IR/DataLayout.h>
+#include <llvm/IR/Instruction.h>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/IR/Constants.h>
+#include <llvm/IR/Type.h>
+#include <llvm/Support/raw_ostream.h>
+#include <llvm/Transforms/IPO/PassManagerBuilder.h>
+#define TRACE_FUN_NAME "trace"
+#define GUARD_FUN_NAME "__guard_failure"
+#define UNOPT_PREFIX "__unopt_"
+
+using namespace llvm;
+
+namespace {
+
+  bool isJump(unsigned);
+
+  /*
+   * Insert a stackmap call before each call to `putchar`.
+   */
+  struct CheckPointPass: public FunctionPass {
+    static char id;
+    /*
+     * The unique identifier of the next stackmap call.
+     */
+    static uint64_t sm_id;
+
+    /*
+     * Map the function name to the ID of the first stackmap/patchpoint call
+     * in the function. This is used to ensure the runtime can associate the
+     * stackmap calls inside each function with stackmap calls from the
+     * unoptimized version of the function
+     */
+    static std::map<StringRef, uint64_t> start_sm_id;
+
+    CheckPointPass() : FunctionPass(id) {}
+
+    virtual bool runOnFunction(Function &fun) {
+      outs() << "Running CheckPointPass on function: " << fun.getName() << '\n';
+      StringRef fun_name = fun.getName();
+      if (!fun_name.endswith(TRACE_FUN_NAME)) {
+        return false;
+      }
+
+      // "fun" -> "__unopt_fun" or
+      // "__unopt_fun" -> "fun"
+      StringRef complementary_fun = "";
+      if (fun_name.startswith(UNOPT_PREFIX)) {
+        complementary_fun = fun_name.substr(StringRef(UNOPT_PREFIX).size());
+      } else {
+        complementary_fun = StringRef(UNOPT_PREFIX + fun_name.str());
+      }
+
+      uint64_t curr_sm_id = sm_id;
+      if (start_sm_id.find(complementary_fun) != start_sm_id.end()) {
+        curr_sm_id = ~start_sm_id[complementary_fun];
+      }
+
+      Module *mod = fun.getParent();
+      DataLayout data_layout(mod);
+      start_sm_id[fun_name] = sm_id;
+      uint64_t call_inst_count = 0;
+      for (auto &bb : fun) {
+        for (BasicBlock::iterator it = bb.begin(); it != bb.end(); ++it) {
+          if (isa<CallInst>(it) &&
+              cast<CallInst>(*it).getCalledFunction()->getName() == "putchar") {
+            LLVMContext &ctx = bb.getContext();
+            IRBuilder<> builder(&bb, it);
+            Type *i8ptr_t = PointerType::getUnqual(IntegerType::getInt8Ty(ctx));
+            Constant* gf_handler_ptr = ConstantExpr::getBitCast(
+                mod->getFunction(GUARD_FUN_NAME), i8ptr_t);
+
+            uint64_t patchpoint_id;
+
+            if (start_sm_id.find(complementary_fun) == start_sm_id.end()) {
+              patchpoint_id = curr_sm_id + call_inst_count;
+            } else {
+              patchpoint_id = ~(~curr_sm_id + call_inst_count);
```

Comment:
What is the `~` about in this file? It's not bitwise negation, is it?

Reply:
It is bitwise negation. It is described in the documentation of the pass. It is a bit of a hack, but it is necessary since I need to be able to associate stackmap records from the unoptimized function with records in the optimized one.